### PR TITLE
Support PyTorch==1.0

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
#514 

- Uses `torch.nn.CTCLoss` instead of `warp-ctc` if `pytorch>=1.0.0`.
- `size_average=True` -> `reduction="mean"` if `pytorch>=1.0.0` or `reduction="elementwise_mean"` if `pytorch<1.0.0` .
- Travis: Tests both `0.4.1` and `1.0.0`.

### CODE REVIEW

**Comment by kamo-naoyuki** [comment]
I got worse results with pytorch=1.0.0, and If `mtlalpha=0.0` and `ctc_weight=0.01`, then became  same as v0.4.1. 
Thus the difference comes from `torch.nn.CTCLoss`, but warp-ctc doesn't work with pytorch=1.0.

## Voxforge, mtlalpha=0.5
- dt
    - pytorch=0.4.1

```
write a CER (or TER) result in exp/tr_it_pytorch_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.txt
|         SPKR                          |         # Snt                 # Wrd         |         Corr                    Sub                    Del                   Ins                    Err                  S.Err         |
|         Sum/Avg                       |         1082                  79133         |         89.2                    5.9                    4.9                   2.9                   13.7                   97.5         |
write a WER result in exp/tr_it_pytorch_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.wrd.txt
|         SPKR                          |         # Snt                  # Wrd         |         Corr                    Sub                     Del                    Ins                    Err                   S.Err         |
|         Sum/Avg                       |         1082                   13235         |         58.1                   37.3                     4.5                    7.2                   49.0                    97.5         |
```

- dt
    - pytorch=1.0.0

```
write a CER (or TER) result in exp/tr_it_pytorch_pt1/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.txt
| SPKR                   | # Snt   # Wrd | Corr     Sub     Del     Ins    Err   S.Err  |
| Sum/Avg                | 1082    79133 | 88.7     6.5     4.9     3.8   15.2    98.6  |
write a WER result in exp/tr_it_pytorch_pt1/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.wrd.txt
| SPKR                   |  # Snt  # Wrd  | Corr      Sub     Del     Ins      Err   S.Err  |
| Sum/Avg                |  1082   13235  | 53.5     41.8     4.7     8.0     54.6    98.6  |
```


- et
    - pytorch=0.4.1

```
write a CER (or TER) result in exp/tr_it_pytorch_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.txt
|         SPKR                          |         # Snt                 # Wrd         |         Corr                    Sub                    Del                   Ins                    Err                  S.Err         |
|         Sum/Avg                       |         1055                  77966         |         89.2                    5.9                    4.9                   2.8                   13.6                   98.7         |
write a WER result in exp/tr_it_pytorch_vggblstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.wrd.txt
|         SPKR                          |         # Snt                  # Wrd         |         Corr                    Sub                     Del                    Ins                    Err                   S.Err         |
|         Sum/Avg                       |         1055                   12990         |         57.8                   37.2                     5.0                    7.4                   49.6                    98.7         |

```

- et
    - pytorch=1.0.0

```
write a CER (or TER) result in exp/tr_it_pytorch_pt1/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.txt
| SPKR                   | # Snt   # Wrd | Corr     Sub     Del     Ins    Err   S.Err  |
| Sum/Avg                | 1055    77966 | 88.7     6.6     4.7     3.6   14.9    99.3  |
write a WER result in exp/tr_it_pytorch_pt1/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.wrd.txt
| SPKR                   |  # Snt  # Wrd  | Corr      Sub     Del     Ins      Err   S.Err  |
| Sum/Avg                |  1055   12990  | 54.0     41.0     5.0     8.4     54.4    99.3  |
```


## AN4, mtlalpha=0.5

- dev
    - pytorch=0.4.1

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt0.4/decode_train_dev_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5/result.txt
|   SPKR     |   # Snt     # Wrd   |   Corr        Sub         Del        Ins        Err      S.Err   |
|   Sum/Avg  |    100       1915   |   83.1        7.1         9.8        1.7       18.6       77.0   |
```

- dev
    - pytorch=1.0.0

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1.0/decode_train_dev_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5/result.txt
|   SPKR     |   # Snt     # Wrd   |   Corr        Sub         Del        Ins        Err      S.Err   |
|   Sum/Avg  |    100       1915   |   56.7       16.0        27.3        2.0       45.3       93.0   |
```

- test
    - pytorch=0.4.1
```
write a CER (or TER) result in exp/train_nodev_pytorch_pt0.4/decode_test_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5/result.txt
|   SPKR     |  # Snt     # Wrd   |  Corr        Sub       Del        Ins       Err      S.Err   |
|   Sum/Avg  |   130       2565   |  90.4        4.4       5.1        0.9      10.5       59.2   |
```

- test
    - pytorch=1.0.0

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1.0/decode_test_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5/result.txt
|   SPKR     |  # Snt     # Wrd   |  Corr        Sub       Del        Ins       Err      S.Err   |
|   Sum/Avg  |   130       2565   |  65.0       13.3      21.7        2.2      37.2       85.4   |
```

## AN4, mtlalpha=0.0, ctc_weight=0.0

- dev
    - pytorch=0.4.1

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt0.4_mtlalpha0.0_ctcw0.0/decode_train_dev_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.0/result.txt
|    SPKR      |    # Snt        # Wrd    |    Corr          Sub           Del          Ins          Err         S.Err    |
|    Sum/Avg   |     100          1915    |    50.4         31.1          18.5         21.0         70.7          86.0    |
```

- dev
    - pytorch=1.0.0

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1_mtlalpha0.0_ctcw0.0/decode_train_dev_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.0/result.txt
|    SPKR      |    # Snt       # Wrd    |    Corr          Sub           Del          Ins          Err        S.Err    |
|    Sum/Avg   |     100         1915    |    50.4         31.1          18.5         21.0         70.7         86.0    |
```

- test
    - pytorch=0.4.1

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt0.4_mtlalpha0.0_ctcw0.0/decode_test_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.0/result.txt
|    SPKR      |    # Snt      # Wrd    |    Corr          Sub          Del         Ins          Err        S.Err    |
|    Sum/Avg   |     130        2565    |    52.5         28.9         18.6        19.4         66.9         86.9    |
```


- test
    - pytorch=1.0.0

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1_mtlalpha0.0_ctcw0.0/decode_test_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.0/result.txt
|    SPKR      |   # Snt       # Wrd    |   Corr          Sub         Del          Ins         Err        S.Err    |
|    Sum/Avg   |    130         2565    |   52.5         28.9        18.6         19.4        66.9         86.9    |
```

**Comment by sw005320** [comment]
Can you run it only with CTC?
We could see the degradation comes from pure CTC or the combination with attention.

**Comment by kamo-naoyuki** [comment]
(I found a small bug only with CTC mode and I'll send PR later)

# AN4, mtlalpha=1.0, ctc_weight=1.0


- dev
    - pytorch=0.4.1
```
write a CER (or TER) result in exp/train_nodev_pytorch_pt0.4_mtlalpha1.0_ctcw1.0/decode_train_dev_beam20_emodel.loss.best_p0.0_len0.0-0.0_ctcw1.0/result.txt
|    SPKR      |    # Snt        # Wrd    |    Corr           Sub          Del           Ins          Err         S.Err    |
|    Sum/Avg   |     100          1915    |    82.7           5.1         12.2           1.6         18.9          82.0    |
```


    - pytorch=1.0.0

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1_mtlalpha1.0_ctcw1.0/decode_train_dev_beam20_emodel.loss.best_p0.0_len0.0-0.0_ctcw1.0/result.txt
|    SPKR      |    # Snt       # Wrd     |    Corr          Sub          Del          Ins           Err        S.Err    |
|    Sum/Avg   |     100         1915     |    61.6          6.9         31.5          0.7          39.1         94.0    |
```

- test
    - pytorch=0.4.1
```
write a CER (or TER) result in exp/train_nodev_pytorch_pt0.4_mtlalpha1.0_ctcw1.0/decode_test_beam20_emodel.loss.best_p0.0_len0.0-0.0_ctcw1.0/result.txt
|    SPKR      |    # Snt       # Wrd    |    Corr         Sub          Del          Ins          Err        S.Err    |
|    Sum/Avg   |     130         2565    |    89.9         3.3          6.8          0.9         11.0         74.6    |
```


    - pytorch=1.0.0

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1_mtlalpha1.0_ctcw1.0/decode_test_beam20_emodel.loss.best_p0.0_len0.0-0.0_ctcw1.0/result.txt
|    SPKR      |   # Snt       # Wrd    |    Corr         Sub          Del          Ins         Err        S.Err    |
|    Sum/Avg   |    130         2565    |    74.5         4.7         20.8          1.2        26.6         93.8    |
```

**Comment by kamo-naoyuki** [comment]
I tried `jnishi/warp-ctc` and `SeanNaren/warp-ctc`, but the former can't be compiled with pytorch=1.0, maybe needs small changes, and the latter exited with segmentation fault.

**Comment by sw005320** [comment]
`jnishi/warp-ctc` just put a small patch for `SeanNaren/warp-ctc`, and if `SeanNaren/warp-ctc` does not support pytorch=1.0, we could not use it.

**Comment by jnishi** [comment]
Use https://github.com/jnishi/warp-ctc/tree/pytorch-1.0.0 with pytorch 1.0.0 because https://github.com/SeanNaren/warp-ctc/pull/89 is merged to the branch. 

I don't know whether this modification fits to ESPNet because error happens when I just apply the PR, but it seems to be fixed by modification of `espnet/nets/pytorch_backend/encoders.py`.

```
exception in main training loop: can't convert CUDA tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
Traceback (most recent call last):
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/chainer/training/trainer.py", line 315, in run
    update()
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/chainer/training/updaters/standard_updater.py", line 165, in update
    self.update_core()
  File "/home/christopher/gitrepos/espnet-1/espnet/asr/pytorch_backend/asr.py", line 141, in update_core
    loss = self.model(*x)[0]
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/christopher/gitrepos/espnet-1/espnet/nets/pytorch_backend/e2e_asr.py", line 183, in forward
    hs_pad, hlens = self.enc(xs_pad, ilens)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/christopher/gitrepos/espnet-1/espnet/nets/pytorch_backend/encoders.py", line 206, in forward
    xs_pad, ilens = module(xs_pad, ilens)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/christopher/gitrepos/espnet-1/espnet/nets/pytorch_backend/encoders.py", line 150, in forward
    np.ceil(np.array(ilens, dtype=np.float32) / 2), dtype=np.int64)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/tensor.py", line 452, in __array__
    return self.numpy().astype(dtype, copy=False)
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "/home/christopher/gitrepos/espnet-1/egs/librispeech/asr1/../../../espnet/bin/asr_train.py", line 232, in <module>
    main(sys.argv[1:])
  File "/home/christopher/gitrepos/espnet-1/egs/librispeech/asr1/../../../espnet/bin/asr_train.py", line 226, in main
    train(args)
  File "/home/christopher/gitrepos/espnet-1/espnet/asr/pytorch_backend/asr.py", line 405, in train
    trainer.run()
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/chainer/training/trainer.py", line 329, in run
    six.reraise(*sys.exc_info())
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/chainer/training/trainer.py", line 315, in run
    update()
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/chainer/training/updaters/standard_updater.py", line 165, in update
    self.update_core()
  File "/home/christopher/gitrepos/espnet-1/espnet/asr/pytorch_backend/asr.py", line 141, in update_core
    loss = self.model(*x)[0]
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/christopher/gitrepos/espnet-1/espnet/nets/pytorch_backend/e2e_asr.py", line 183, in forward
    hs_pad, hlens = self.enc(xs_pad, ilens)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/christopher/gitrepos/espnet-1/espnet/nets/pytorch_backend/encoders.py", line 206, in forward
    xs_pad, ilens = module(xs_pad, ilens)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/christopher/gitrepos/espnet-1/espnet/nets/pytorch_backend/encoders.py", line 150, in forward
    np.ceil(np.array(ilens, dtype=np.float32) / 2), dtype=np.int64)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python3.6/site-packages/torch/tensor.py", line 452, in __array__
    return self.numpy().astype(dtype, copy=False)
TypeError: can't convert CUDA tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

**Comment by kamo-naoyuki** [comment]
@jnishi Thanks.

# AN4, mtlalpha=0.5, ctc_weight=0.5, pytorch=1.0 with https://github.com/jnishi/warp-ctc/tree/pytorch-1.0.0 

Got better result than 0.4.1. I don't know why.
(I also check with mtlalpha=1.0 again and got same results between 1.0.0 and 0.4.1 )

```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1.0_mtlalpha0.5_ctcw0.5/decode_train_dev_beam20_emodel.loss.best_p0.0_len0.0-0.0_ctcw0.5/result.txt
|    SPKR      |    # Snt        # Wrd    |    Corr           Sub          Del           Ins          Err         S.Err    |
|    Sum/Avg   |     100          1915    |    84.4           5.4         10.1           1.5         17.1          70.0    |

```
```
write a CER (or TER) result in exp/train_nodev_pytorch_pt1.0_mtlalpha0.5_ctcw0.5/decode_test_beam20_emodel.loss.best_p0.0_len0.0-0.0_ctcw0.5/result.txt
|    SPKR      |    # Snt       # Wrd    |    Corr         Sub          Del          Ins          Err        S.Err    |
|    Sum/Avg   |     130         2565    |    91.2         3.6          5.2          0.7          9.5         56.2    |
```

**Comment by sw005320** [comment]
> Got better result than 0.4.1. I don't know why.
(I also check with mtlalpha=1.0 again and got same results between 1.0.0 and 0.4.1 )

This is great. AN4 is too small to discuss details about why, but anyway, it's better to have better results.
Can you also test it with the voxforge recipe?

**Comment by kamo-naoyuki** [comment]
Slightly better result than 0.4.1 with Voxforge, too.
Hmm, was warp-ctc also changed anything recently? Does anyone know?

## Voxforge(it), mtlalpha=0.5, ctc_weight=0.5, pytorch=1.0


- dt
    - pytorch=0.4.1 (warp-ctc)

```
write a CER (or TER) result in exp/tr_it_pytorch_pt0.4_mtlalpha0.5/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.txt
|  SPKR                   |  # Snt     # Wrd  |  Corr      Sub       Del      Ins      Err     S.Err  |
|  Sum/Avg                |  1082      79133  |  89.1      5.9       5.1      2.9     13.8      97.9  |
write a WER result in exp/tr_it_pytorch_pt0.4_mtlalpha0.5/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.wrd.txt
|  SPKR                    |  # Snt    # Wrd   |  Corr       Sub      Del       Ins       Err     S.Err   |
|  Sum/Avg                 |  1082     13235   |  57.0      38.4      4.5       7.1      50.1      97.9   |

```

- dt
    - pytorch=1.0.0 (warp-ctc)

```
write a CER (or TER) result in exp/tr_it_pytorch_pt1.0_mtlalpha0.5/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.txt
|  SPKR                   |  # Snt     # Wrd  |  Corr      Sub       Del      Ins      Err     S.Err  |
|  Sum/Avg                |  1082      79133  |  89.6      5.8       4.6      2.9     13.3      98.2  |
write a WER result in exp/tr_it_pytorch_pt1.0_mtlalpha0.5/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.wrd.txt
|  SPKR                    |  # Snt    # Wrd   |  Corr       Sub      Del       Ins       Err     S.Err   |
|  Sum/Avg                 |  1082     13235   |  57.3      37.6      5.1       5.8      48.6      98.2   |
```

- dt
    - pytorch=1.0.0 (torch.nn.CTCLoss)

```
write a CER (or TER) result in exp/tr_it_pytorch_pt1/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.txt
| SPKR                   | # Snt   # Wrd | Corr     Sub     Del     Ins    Err   S.Err  |
| Sum/Avg                | 1082    79133 | 88.7     6.5     4.9     3.8   15.2    98.6  |
write a WER result in exp/tr_it_pytorch_pt1/decode_dt_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.wrd.txt
| SPKR                   |  # Snt  # Wrd  | Corr      Sub     Del     Ins      Err   S.Err  |
| Sum/Avg                |  1082   13235  | 53.5     41.8     4.7     8.0     54.6    98.6  |
```


- et
    - pytorch=0.4.1 (warp-ctc)

```
write a CER (or TER) result in exp/tr_it_pytorch_pt0.4_mtlalpha0.5/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.txt
|  SPKR                   |  # Snt     # Wrd  |  Corr      Sub       Del      Ins      Err     S.Err  |
|  Sum/Avg                |  1055      77966  |  89.0      6.0       5.0      2.6     13.6      98.7  |
write a WER result in exp/tr_it_pytorch_pt0.4_mtlalpha0.5/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.wrd.txt
|  SPKR                    |  # Snt    # Wrd   |  Corr       Sub      Del       Ins       Err     S.Err   |
|  Sum/Avg                 |  1055     12990   |  56.7      38.2      5.1       7.4      50.7      98.7   |

```

- et
    - pytorch=1.0.0 (warp-ctc)

```
write a CER (or TER) result in exp/tr_it_pytorch_pt1.0_mtlalpha0.5/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.txt
|  SPKR                   |  # Snt     # Wrd  |  Corr      Sub       Del      Ins      Err     S.Err  |
|  Sum/Avg                |  1055      77966  |  89.6      5.7       4.7      2.6     13.0      99.1  |
write a WER result in exp/tr_it_pytorch_pt1.0_mtlalpha0.5/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.5/result.wrd.txt
|  SPKR                    |  # Snt    # Wrd   |  Corr       Sub      Del       Ins       Err     S.Err   |
|  Sum/Avg                 |  1055     12990   |  56.8      37.4      5.8       5.8      49.0      99.1   |

```

- et
    - pytorch=1.0.0( torch.nn.CTCLoss)

```
write a CER (or TER) result in exp/tr_it_pytorch_pt1/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.txt
| SPKR                   | # Snt   # Wrd | Corr     Sub     Del     Ins    Err   S.Err  |
| Sum/Avg                | 1055    77966 | 88.7     6.6     4.7     3.6   14.9    99.3  |
write a WER result in exp/tr_it_pytorch_pt1/decode_et_it_beam20_emodel.acc.best_p0_len0.0-0.0_ctcw0.3/result.wrd.txt
| SPKR                   |  # Snt  # Wrd  | Corr      Sub     Del     Ins      Err   S.Err  |
| Sum/Avg                |  1055   12990  | 54.0     41.0     5.0     8.4     54.4    99.3  |
```

**Comment by kamo-naoyuki** [comment]
About current setting of `Makefile`:

- Both `1.0.0` and `0.4.1` are supported and sets pytorch=0.4.1 as default, for the time being.
    - It can be changed by `TH_VERSION` of Makefile.
-  Keeping the usage of  both `Warp-ctc` and `torch.nn.CTCLoss`, and uses `Warp-ctc` by default.
- `GCC>=4.9` is required for v1.0.0 only when compiling `Warp-ctc`. This is an official requirement of pytorch=1.0.0.

I haven't tested `tts` mode. I don't know how to check the result.

**Comment by kan-bayashi** [comment]
There is no quantitative result for tts so you can ignore.

**Comment by kamo-naoyuki** [comment]
Thanks, then it's O.K to merge for me. Please review.

**Comment by sw005320** [comment]
For now, we still keep to use WARP_CTC instead of switch it with an option, right?https://github.com/espnet/espnet/pull/553/files#diff-f01ba30066efe1140636f8abffc716bbR10

This is fine, but note that the chainer backend has an option to switch to use warpctc or chaner ctc in https://github.com/espnet/espnet/pull/553/files#diff-f01ba30066efe1140636f8abffc716bbR10
You may also think about switching it with an option.

**Comment by kamo-naoyuki** [comment]
@sw005320 O.K. I'll take in it.

https://github.com/espnet/espnet/blob/fff3d08bac3216162c63373c0d09483f18d8f281/espnet/bin/asr_train.py#L63-L65

The arguments of asr_train.py are common between chainer and pytorch, so I think this "chainer"  should be changed with some word,  how about "built-in"? 

Or,  changes as following,

```python
choices=['chainer', 'warpctc', 'pytorch'], 
```

and if "chainer" is choosed with pytorch backend, `ctc_for` raises `RuntimeError` , and vice versa.

Which do you like?

**Comment by sw005320** [comment]
> The arguments of asr_train.py are common between chainer and pytorch, so I think this "chainer" should be changed with some word, how about "built-in"?

"built-in" sounds good.

**Comment by kamo-naoyuki** [comment]
or "native" is better than "built-in"? I'm not sure.

File: tools/Makefile [hunk 1]
```diff
@@ -1,5 +1,7 @@
 PYTHON := /usr/bin/python2.7
+TH_VERSION := 0.4.1
```
[the remark this answers is not in the document]
Comment (kamo-naoyuki) [reply tools/Makefile]: It's ok for me, too, if you don't mind.

**Pull request merged**